### PR TITLE
feat(app-platform): Sort Integrations

### PR DESCRIPTION
**Pull request by mnoble**
"Sentry Apps" and Integrations are now sorted among one another. First by Installed vs Uninstalled status, then alphabetically within each of those groups.

### CODE REVIEW

File: src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx [hunk 1]
```diff
@@ -169,21 +215,17 @@ class OrganizationIntegrations extends AsyncComponent {
...
+            {installed}
+            {uninstalled}
           </PanelBody>
         </Panel>
       </React.Fragment>
     );
   }
 }
 
+const IntegrationRow = styled.div``;
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx]: Do we need this component?

**Review by markstory — COMMENTED** [review]
I'm not sure those massive snapshots are providing much value. You might want to consider using a visual snapshot or an acceptance test instead.

File: src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx (hunk 1 above)
Comment (mnoble) [reply src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx]: Functionally, no, but SentryApps and Integrations use two different components for each row. This makes testing the sorting a lot easier (find all `IntegrationRow` items, regardless of which type they are). Long term, I plan to consolidate these two sets of components into one unified set, but that's down the line.

Comment (markstory) [reply src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx]: I more meant that this is an empty div with no styling. You could just `<div>` for now.

Comment (billyvg) [reply src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx]: They're using the component name as a selector in tests (which I find a bit more pleasant to read)

File: src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx [hunk 2]
```diff
@@ -148,7 +146,55 @@ class OrganizationIntegrations extends AsyncComponent {
...
+          applications={apps}
+        />
+      </IntegrationRow>
+    );
+  }
+
+  renderBody() {
+    const {reloading, applications, appInstalls} = this.state;
+
+    const installedProviders = this.providers
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx]: There's kind of a lot of duplicated and/or very similar code, we could use `lodash.partition` to cut back some code. Downside of this is that readers will need to look up what `partition` does.  (Current code is fine too though).

```
const [installedProviders, uninstalledProviders] = _.partition(this.providers.map(p => [p, this.renderProvider(p)]), ([provider,]) => provider.isInstalled);
```

File: src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx [hunk 3]
```diff
@@ -148,7 +146,55 @@ class OrganizationIntegrations extends AsyncComponent {
...
+      .map(a => [a.name, this.renderSentryApps([a], a.slug)]);
+
+    const uninstalledSentryApps = (applications || [])
+      .filter(a => !appInstalls.find(i => i.app.slug === a.slug))
+      .map(a => [a.name, this.renderSentryApps([a], a.slug)]);
+
+    // Combine the list of Providers and Sentry Apps that have installations.
+    const installed = installedProviders
+      .concat(installedSentryApps)
+      .sort((a, b) => a[0].localeCompare(b[0]))
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx]: imo, destructuring adds some more code but can be a bit more clear about what you're retrieving from the array:

```
const installed = installedProviders
      .concat(installedSentryApps)
      .sort(([name1,], [name2,]) => name1.localeCompare(name2))
      .map([,Component] => Component);
```

**Review by billyvg — APPROVED** [review]
Looks good, we should figure out better tests for `tests/js/spec/views/settings/organizationIntegrations/__snapshots__/index.spec.jsx.snap` instead of a massive 10k line snapshot

**Comment by mnoble** [comment]
@billyvg 👍  Sorta tangential, but what's the benefit of those snapshot tests? I don't think I ever really understood. I suppose if you're changing data only, they ensure the rendering is unaffected? Seems like most of the time you're changing rendering and are just going to re-snapshot it, though.

**Comment by billyvg** [comment]
@mnoble when they are 10k lines long... none and I think overall they provide little value.

**Comment by mnoble** [comment]
Heh, fair enough.

**Comment by markstory** [comment]
@mnoble Personally I would trash those jest snapshots and add percy snapshots and an acceptance test.

File: src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx (hunk 3 above)
Comment (mnoble) [reply src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx]: Yea, more "punctuation", but definitely clearer 👍

**Pull request merged**